docs/AUTH_PROTOCOL.md

```diff
@@ -0,0 +1,393 @@
+
+The rest of the flow's information is for display purposes _only_. The flow's `primaryMechanism` is used to determine how the flow will be presented to the user: which icons to use, and which name to display (including localization). This makes basic flows very convenient: just send `{ "id": "password", "primaryMechanism": "password" }` and the frontend will take care of displaying the flow optimally to the user. However, for more custom use-cases where you have some custom flow, the flow may also contain a `name` that will override the name selected via `primaryMechanism`. Note that the flow is not limited to only using the `primaryMechanism`.
+
+The frontend will either prompt the user to select a flow, or pre-select a flow for the user. Once a flow is selected, the frontend communicates this to the authenticator:
+
+```js
+{
+    "event": "start",
+    "flow": "mycorp-login-flow",
+}
+```
+
+From this point forward, the protocol becomes a lot more event-driven. The authenticator will start sending events that request the display of certain UI components, and the frontend communicates the user's input back to the authenticator.
+
+The frontend may want to restart a flow (i.e. if the user hits the "back" button) or switch flows (i.e. if the user selects a different flow from a menu). Both can be achieved by sending the `start` event a second time with the cooresponding flow ID.
+
+## Authentication Mechanisms and Conversation
+
+Authentication mechanisms are standardized UI patterns that an authenticator can request from a frontend. For instance: if the authenticator asks the frontend for a password, the frontend will pick an appropriate UI for password entry. Mechanisms can be categorized as "primary" or "secondary". There can only be one primary mechanism active (i.e. visible on screen) at a time, so if the authenticator asks to show a different primary mechanism the previous UI should be replaced with the new one. Secondary mechanisms can co-exist with the primary mechanism and with each-other
+
+Once a flow is started, the authentication "conversation" begins. At a high level, a conversation looks like the following:
+
+1. The authenticator requests a primary authentication mechanism from the frontend
+2. The frontend displays the appropriate UI for this authentication mechanism
+3. The user interacts with the UI to give it some data (types in a password, touches a fingerprint sensor, etc)
+4. Once the user is finished, the frontend sends the user's entered data to the authenticator
+5. If the user has failed to authenticate, or if there are no more steps left in the flow, the authenticator ends the conversation and returns success or failure
+6. Otherwise, the authenticator continues with the next step of the flow. Return to step 1
```

Comment:
When requesting a new authentication mechanism from the frontend, the authenticator should have the option to show a message to the user. 

For example, authd refreshes the access token during local password authentication. When that fails for some reason (e.g. the refresh token being expired), it automatically starts device authentication and shows the user an explanatory message.

Reply:
Makes sense. Plus, a generic "message" mechanism that does nothing but display some text might be useful too